src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs

```diff
@@ -0,0 +1,101 @@
+                                                , lookupSmartIdent
+                                                )
+import           FreeC.Environment.Fresh        ( freshAgdaVar )
+import           FreeC.Monad.Converter          ( Converter
+                                                , localEnv
+                                                )
+
+
+-- | Converts a data or type synonym declaration.
+convertTypeDecl :: IR.TypeDecl -> Converter [Agda.Declaration]
+convertTypeDecl (IR.TypeSynDecl _ _ _ _) = error "Not supported at the moment."
+convertTypeDecl (IR.DataDecl _ ident tVars constrs) =
+  (:) <$> convertDataDecl ident tVars constrs <*> convertSmartConstrs constrs
+
+-- | Converts a data declaration.
+convertDataDecl
+  :: IR.DeclIdent
+  -> [IR.TypeVarDecl]
+  -> [IR.ConDecl]
+  -> Converter Agda.Declaration
+convertDataDecl ident tVars constrs =
+  localEnv
+    $   freeDataDecl
+    <$> lookupTypeIdent ident
+    <*> mapM convertTypeVarDecl tVars
+    <*> convertConstructors ident tVars constrs
+
+-- | Converts all constructors of a data declaration.
+--
+--   We synthesize the return type of the constructor in IR to avoid passing the
+--   identifier and all used type variables to the type converter. This way we
+--   can reuse the existing translation.
+convertConstructors
+  :: IR.DeclIdent     -- ^ The identifier of the data type.
+  -> [IR.TypeVarDecl] -- ^ The type parameters declared by the data type.
+  -> [IR.ConDecl]     -- ^ The constructor declarations of the data type.
+  -> Converter [Agda.Declaration]
+convertConstructors (IR.DeclIdent srcSpan ident) tVars =
+  mapM $ convertConstructor $ IR.typeApp IR.NoSrcSpan
+                                         (IR.TypeCon srcSpan ident)
+                                         (map IR.typeVarDeclToType tVars)
+
+-- | Converts a single constructor of a (non-recursive) data type.
+convertConstructor :: IR.Type -> IR.ConDecl -> Converter Agda.Declaration
+convertConstructor retType (IR.ConDecl _ ident argTs) =
+  Agda.funcSig
+    <$> lookupValueIdent ident
+    <*> convertConstructorType argTs retType
+
+-- | Converts a list of constructors to a list of smart constructors.
+convertSmartConstrs :: [IR.ConDecl] -> Converter [Agda.Declaration]
+convertSmartConstrs = mapM convertSmartConstr
+
+-- | Converts a single constructor to a smart constructor, which wraps the normal
+--   constructor in the @Free@ monad using @pure@.
+--
+--   We use Agda's @pattern@ declarations for smart constructors to simplify
+--   pattern matching in proofs.
+convertSmartConstr :: IR.ConDecl -> Converter Agda.Declaration
+convertSmartConstr (IR.ConDecl _ ident argTs) = do
+  smartName <- lookupSmartIdent ident
+  patternDecl smartName (repeat "x" `zip` argTs) $ \vars -> do
+    normalName <- Agda.IdentP . Agda.qname' <$> lookupValueIdent ident
+    let pureVal = foldl Agda.appP normalName vars
+    pure (Agda.IdentP (Agda.qname' Agda.Base.pure) `Agda.appP` pureVal)
```

Comment:
Try to use `return` instead of `pure` (especially if you are in a `do` block). I would use `pure` only if there is no `Monad` instance.